### PR TITLE
Fixed #31636 -- Made BooleanFieldListFilter respect Field.choices.

### PR DESCRIPTION
**Pull request by manav014**
Change the behaviour of BooleanFieldListFilter in such a way that choices are
respected while populating filters on the admin page.

[ticket-31636](https://code.djangoproject.com/ticket/31636)

### CODE REVIEW

File: django/contrib/admin/filters.py [hunk 1]
```diff
@@ -236,18 +236,26 @@ def __init__(self, field, request, params, model, model_admin, field_path):
...
+        else:
+            list_of_choices = [
                 ('1', _('Yes')),
-                ('0', _('No'))):
+                ('0', _('No')),
+            ]
+        for lookup, title in (
+            (None, _('All')),
+            *list_of_choices
+        ):
```
Comment (felixxm) [inline django/contrib/admin/filters.py:258]: We should respect also description for `None` (test is required).

File: django/contrib/admin/filters.py [hunk 2]
```diff
@@ -236,18 +236,26 @@ def __init__(self, field, request, params, model, model_admin, field_path):
...
-        for lookup, title in (
-                (None, _('All')),
+        if self.field.flatchoices:
+            list_of_choices = [(str(int(x[0])), x[1]) for x in self.field.flatchoices if x[0] is not None]
+        else:
+            list_of_choices = [
                 ('1', _('Yes')),
-                ('0', _('No'))):
+                ('0', _('No')),
+            ]
```
Comment (felixxm) [inline django/contrib/admin/filters.py:254]: IMO we could simplify this with using `dict`, e.g.
```python
def choices(self, changelist):
    choices = {
        None: _('All'),
        True: _('Yes'),
        False: _('No'),
        **{lookup: title for lookup, title in self.field.flatchoices},
    }
    for lookup, title in choices.items():
        lookup_val = str(int(lookup)) if lookup is not None else lookup
        yield {
            'selected': self.lookup_val == lookup_val and not self.lookup_val2,
            'query_string': changelist.get_query_string({self.lookup_kwarg: lookup_val}, [self.lookup_kwarg2]),
            'display': title,
        }
    ...
```

File: django/contrib/admin/filters.py [hunk 3]
```diff
@@ -315,7 +323,7 @@ def __init__(self, field, request, params, model, model_admin, field_path):
 
         if isinstance(field, models.DateTimeField):
             today = now.replace(hour=0, minute=0, second=0, microsecond=0)
-        else:       # field is a models.DateField
+        else:  # field is a models.DateField
```
Comment (felixxm) [inline django/contrib/admin/filters.py:326]: Please revert this unrelated change.

File: tests/admin_filters/models.py [hunk 4]
```diff
@@ -31,6 +31,11 @@ class Book(models.Model):
     is_best_seller = models.BooleanField(default=0, null=True)
     is_best_seller2 = models.NullBooleanField(default=0)
     date_registered = models.DateField(null=True)
+    availability_choices = (
+        (False, 'paid'),
+        (True, 'free'),
+    )
+    availability = models.BooleanField(choices=availability_choices, null=True)
```
Comment (felixxm) [inline tests/admin_filters/models.py:38]: We don't reuse `availability_choices` anywhere so there is no need for a separate variable.

File: tests/admin_filters/tests.py [hunk 5]
```diff
@@ -140,12 +140,24 @@ class CustomUserAdmin(UserAdmin):
 
 
 class BookAdmin(ModelAdmin):
-    list_filter = ('year', 'author', 'contributors', 'is_best_seller', 'date_registered', 'no')
+    list_filter = ('year', 'author', 'contributors', 'is_best_seller', 'date_registered', 'no', 'availability')
```
Comment (felixxm) [inline tests/admin_filters/tests.py:143]: This change is not required, please revert it.

File: tests/admin_filters/tests.py [hunk 6]
```diff
@@ -140,12 +140,24 @@ class CustomUserAdmin(UserAdmin):
...
     ordering = ('-id',)
 
 
 class BookAdmin2(ModelAdmin):
-    list_filter = ('year', 'author', 'contributors', 'is_best_seller2', 'date_registered', 'no')
+    list_filter = ('year', 'author', 'contributors', 'is_best_seller2', 'date_registered', 'no', 'availability')
```
Comment (felixxm) [inline tests/admin_filters/tests.py:148]: This change is not required, please revert it.

File: tests/admin_filters/tests.py [hunk 7]
```diff
@@ -140,12 +140,24 @@ class CustomUserAdmin(UserAdmin):
...
+class BookAdminWithChoicesBooleanFilter(BookAdmin):
+    list_filter = (
+        'year',
+        'author',
+        'contributors',
+        'is_best_seller',
+        'date_registered',
+        'no',
+        ('availability', BooleanFieldListFilter),
+    )
```
Comment (felixxm) [inline tests/admin_filters/tests.py:160]: I would add `('availability', BooleanFieldListFilter)` to the `BookAdminWithTupleBooleanFilter` instead of creating a separate class.

**Review by felixxm — COMMENTED** [review]
@manav014 Thanks for this patch :+1: I left comments.

File: django/contrib/admin/filters.py [hunk 8]
```diff
@@ -236,18 +236,26 @@ def __init__(self, field, request, params, model, model_admin, field_path):
         self.lookup_val = params.get(self.lookup_kwarg)
         self.lookup_val2 = params.get(self.lookup_kwarg2)
         super().__init__(field, request, params, model, model_admin, field_path)
-        if (self.used_parameters and self.lookup_kwarg in self.used_parameters and
+        if (self.used_parameters and
+                self.lookup_kwarg in self.used_parameters and
```
[the remark this answers is not in the document]
Comment (manav014) [reply django/contrib/admin/filters.py]: Reverted the unrelated change.

File: django/contrib/admin/filters.py (hunk 3 above)
Comment (manav014) [reply django/contrib/admin/filters.py]: Reverted this unrelated change.

File: tests/admin_filters/tests.py (hunk 5 above)
Comment (manav014) [reply tests/admin_filters/tests.py:143]: Reverted this unrelated change.

File: django/contrib/admin/filters.py (hunk 1 above)
Comment (manav014) [reply django/contrib/admin/filters.py]: Added the test for None as well

File: tests/admin_filters/tests.py (hunk 6 above)
Comment (manav014) [reply tests/admin_filters/tests.py]: Reverted this unrelated change.

File: tests/admin_filters/models.py (hunk 4 above)
Comment (manav014) [reply tests/admin_filters/models.py]: Removed the separate variable for the same.

File: django/contrib/admin/filters.py (hunk 2 above)
Comment (manav014) [reply django/contrib/admin/filters.py]: Simplified this using `dict`

File: tests/admin_filters/tests.py (hunk 7 above)
Comment (manav014) [reply tests/admin_filters/tests.py]: Removed the `BookAdminWithChoicesBooleanFilter` and added `('availability', BooleanFieldListFilter)`to the `BookAdminWithTupleBooleanFilter` .

**Review by felixxm — COMMENTED** [review]
@manav014 Thanks :+1: Welcome aboard :boat: 

I added tests for `False` and `None`.
